docs/proposals/004-initialization.md

```diff
@@ -0,0 +1,317 @@
+SP #004: Initialization
+=================
+
+This proposal documents the desired behavior of initialization related language semantics, including default constructor, initialization list and variable initialization.
+
+Status
+------
+
+Status: Design Review
+
+Implemtation: N/A
+
+Author: Yong He
+
+Reviewer: Theresa Foley, Kai Zhang
+
+Background
+----------
+
+Slang has introduced several different syntax around initialization to provide syntactic compatibility with HLSL/C++. As the language evolve, there are many corners where
+the semantics around initialization are not well-defined, and causing confusion or leading to surprising behaviors.
+
+This proposal attempts to provide a design on where we want the language to be in terms of how initialization is handled in all different places.
+
+Related Work
+------------
+
+C++ has many different ways and syntax to initialize an object: through explicit constructor calls, initialization list, or implicitly in a member/variable declaration.
+A variable in C++ can also be in an uninitialized state after its declaration. HLSL inherits most of these behvior from C++ by allowing variables to be uninitialized.
+
+On the other hand, languages like C# and Swift has a set of well defined rules to ensure every variable is initialized after its declaration.
+C++ allows using the initilization list syntax to initialize an object. The semantics of initialization lists depends on whether or not explicit constructors
+are defined on the type.
+
+Proposed Approach
+-----------------
+
+In this section, we document all concepts and rules related to initialization, constructors and initialization lists.
+
+### Default Initializable type
+
+A type is considered "default-initializable" if it provides a constructor that can take 0 arguments, so that it can be constructed with `T()`.
+
+### Variable Initialization
+
+Generally, a variable is considered uninitialized at its declaration site without an explicit value expression.
+For example,
+```csharp
+void foo()
+{
+  MyType t; // t is considered uninitialized.
```

Comment:
Can you clarify that there is no `__init()` defined for `MyType`?

Reply:
The general principle is never to initialize `t`. The fact that we still need to initialize `t` if `MyType` has default `__init` is just for backward compatbility, and we may in the future no longer do so. If the user is defining this as modern syntax such as `var t : MyType`, we will never initialize whether or not `MyType` has `__init`.

Reply:
So: if `MyType` has a default `__init()`, then:
1. `MyType s;` will default initialize now, but we need to think about a migration path so we don't default initilaize it in the future.
2. `var s : MyType` will not initialize.
3. In a modern module, i.e. a module defined with `module moduleName;`, `MyType s;` will not default initialize.

Reply:
Yes, I understand the rules here.

But what I'm trying to suggest is to change this example as
```
struct MyType1 {
  int x;
}
void foo() {
  MyType t1; // `t1` is initialized with a call to `__init`.
}
```

So we know why it's different from the example in line 59-65. Because there is no default `__init()` defined in `MyType1`.

Reply:
Why? MyType1 shouldn't default initialize because it doesn't even have a user defined ctor.